### ISSUE TITLE
Add a Graph data structure to Godot.

### ISSUE THREAD

**Issue by godotgamedeveloper** [godotengine/godot-proposals#3848]
### Describe the project you are working on

Just reopening, in the GIP tracker, an old proposal that was closed in the main tracker with the migration to the GIP tracker.

I have conducted a search in the GIP tracker and did not see any similar proposal listed.  The old proposal is at:
https://github.com/godotengine/godot/issues/15647

### Describe the problem or limitation you are having in your project

As noted in the old proposal, there can sometimes be a need to get access to a graph or tree data structure in the script code without having to rely on Control nodes like Tree or GraphNode and without having to rely on Astar class.  Astar is use-case oriented in that it is basically a graph for the sole purpose of pathfinding,

A graph data structure can have more uses than the Astar class, such as when no pathfinding is necessary.  Furthermore, using Control nodes for this might be more expensive than rolling a custom implementation in GDScript.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

As noted in the old proposal, Godot could benefit from providing a Reference object type that gives access to a labeled vertex, directed and labeled edge, graph data structure.

The advantages of exposing this to scripting could be pretty awesome for making data structures in projects.  You could create scripts that use it as a graph or convert it into a binary search tree, a splay tree, or whatever you want.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

It can be like how TreeItems work with the Tree, we could have a DataGraph (Reference, facade), DataGraphVertex (Object 0+ labels & a Variant meta), and DataGraphEdge (Object with 1 label & a Variant meta).  This format would keep it as flexible as possible.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Rolling a custom implementation of a graph data structure in GDScript would involve more than just a few lines of code.

### Is there a reason why this should be core and not an add-on in the asset library?

As for the placement, how about we create a module that is dedicated to providing generic data structure and algorithm extensions to the core?  That would open things up for successive additions of a similar nature and give the DataGraph a comfortable place to live outside of core.

For instance, we can create a module specifically for "data_structures" and having a subdirectory for a DataGraph class and its associated classes.

**Comment by fire** [comment]
What game use are you using for your project?

**Comment by godotgamedeveloper** [comment]
@fire I've updated the writeup to indicate game uses.  I've also asked the author of the original proposal to chime in.  Thanks!

**Comment by Zireael07** [comment]
Procedural city here, I end up doing the graph by hand just to be able to tell which intersection connects to which (for mapgen purposes, not actual pathfinding). So there are definite game uses

**Comment by fire** [comment]
@godotgamedeveloper I recommend prototyping a resource api and proposing the api here.

**Comment by KoBeWi** [comment]
> Rolling a custom implementation of a graph data structure in GDScript would involve more than just a few lines of code.

Isn't graph just
```GDScript
class GNode:
    var name_maybe_idk: String

class Edge
    var node1: GNode
    var node2: GNode
    var weight_etc: float

var nodes: Array
var edges: Array
```
?

Plus some methods, but that depends on what you need. Could definitely be a plugin.

**Comment by willnationsdev** [comment]
Iirc, the original proposal ultimately did not go anywhere because all general purpose solutions can easily be made via script/plugins and all performance-optimized graphs have to be tailored for their specific use cases with custom implementations. Thus, making a low-level implementation in core that can be generally used for all games isn't really conceivable.

**Comment by Zireael07** [comment]
Considering the Astar implementation, which is already in core, needs some sort of a graph too, it makes more sense to expose that graph to scripts instead of requiring plugins?

**Comment by Xrayez** [comment]
> As for the placement, how about we create a module that is dedicated to providing generic data structure and algorithm extensions to the core?

Such a module exists, it's called [Goost](https://goostengine.github.io/), and I'm a maintainer/creator of it. 😉

This is certainly a feature worth to be implemented (added to the list of approved proposals in goostengine/goost#7 now), even if it won't cover *all* possible use cases.

For very basic single path graphs, you can also use a [`LinkedList`](https://goost.readthedocs.io/en/latest/classes/class_linkedlist.html) currently implemented in Goost. See also relevant Godot proposal:

- #1522

---

> Plus some methods, but that depends on what you need. Could definitely be a plugin.

The methods that you omit are the actual meat behind graph data structures, more that any other data structure that you could stumble upon, in my experience.

I've done some procedural generation myself using single path graphs (racing game), generated [minimum spanning tree](https://en.wikipedia.org/wiki/Minimum_spanning_tree) for generating caves and so on (in GDScript). Those algorithms are not really specific, but of course could easily become specific.

> Thus, making a low-level implementation in core that can be generally used for all games isn't really conceivable.

If we come up with an implementation which could be useful for at least 30-40% of use cases, that's a win in my book! But of course, that goes against [Godot's development philosophy](https://goost.readthedocs.io/en/latest/development/godot_development_philosophy.html) somewhat. Therefore, implementing a data structure like this is likely better implemented in https://github.com/goostengine/goost, but of course that's up to Godot core developers to decide. At any rate, I think it's better that we first implement such a data structure in Goost as an [experimental feature](https://github.com/goostengine/goost/issues/95) first, see if it works in practice, and then let Godot core developers reconsider this proposal. 🙂

But before starting the work, this actually requires quite serious design work. I'd study existing Graph APIs and come up with an API suitable for solving procedural generation specifically.

**Comment by sairam4123** [comment]
As a city building game developer, I had to roll out my own implementation of graph system. But I made it as similar as Godot's AStar Graph implementation, (except some features that very specific to my game). You can find it [here](https://github.com/OpenSkyCity/Godot-Road-Builder-3D/blob/material-overlay/roads_system/Graph.gd).

I think that just exposing the underlying graph data structure of the AStar should be good enough for most purposes.

**Comment by Xrayez** [comment]
The notion of vertices and edges are implementation details of AStar itself. The API previously proposed in godotengine/godot#15647 looks like a collection of separate classes that extend `Reference` for a graph, and then `Object` for both vertices and edges. I'm not sure what kind of approach would be best to take, but all of my implementations for a graph involved proper separation. Data structures like `LinkedList` operate exactly this way.

I'm leaning towards using separate classes for graph/vertex/edge as well. The most prominent advantage that I could think of right now is that it's much easier to navigate the graph:

- With AStar-like API, you basically have to use `Graph.get_points_connections()` which returns the IDs of neighborhood vertices, and then you have to query the attribute of a vertex that you're interested in using those IDs, does not feel intuitive either.
- With separate classes, you just call `GraphVertex.get_neighborhood()` returning an array containing `GraphVertex` instances. An attribute simply becomes a member of that class.

By the way, [`VisualScript`](https://docs.godotengine.org/en/3.4/classes/class_visualscript.html) can also be seen as a graph, the API looks quite reminiscent and also leans towards distinguishing graph and its vertices via `VisualScriptNode` derived classes.

**Comment by Xrayez** [comment]
By the way, when people try to use AStar for non-pathfinding purposes as a general-purpose graph or try to mimic it in some way, this already signifies that we need a dedicated solution (not necessarily a general-purpose one, if that's a concern).

Here's to illustrate my point. I recall there was a discussion for decoupling state machine logic into a general-purpose `StateMachine` class (see godotengine/godot#25021). I think that users should not be tempted or forced into using classes that are meant to solve specific set of use cases, because that could potentially lead to the bloat of those classes, which goes against [Godot's principles of remaining small](https://docs.godotengine.org/en/stable/about/faq.html#why-does-godot-aim-to-keep-its-core-feature-set-small). I realize that there are many ways to implement a  `StateMachine`, so this still leaves the blind spot so to speak (see goostengine/goost#96), but again, it *does not* mean that there's *no useful general purpose solution*.

In fact, when I stumble upon some general-purpose solution (like, a library which handles various degeneracies in input), I feel good because I no longer need to worry about validating or fixing my input, Of course, this almost always has a cost of a lower performance, but again, based on *specific* use case, this is could be acceptable.

**Comment by Xrayez** [comment]
Should a graph allow self-loops and multiple edges? Some call it a *pseudograph*, but some recognize it as a *general graph*, and call a graph without self-loops and multiple edges just as a *simple graph*.

Whatever the naming convention, does it make sense to impose some kind of constraints to a graph, or should it be users' responsibility?

I think that a general-purpose graph should be allowed to have any kind of structure in my opinion, but maybe we could add several properties like `allow_self_loops` and `allow_multi_edge`. Or just add methods like `find_self_loops()` and `find_multi_edges()` that you could validate yourself (since there are also *loopless graphs* that have multi-edges but no self-loops).

**Comment by Zireael07** [comment]
> I think that a general-purpose graph should be allowed to have any kind of structure in my opinion, but maybe we could add several properties

This is the way to go imho. Properties (or methods) that allow optional specialized things, but the core should be general.

**Comment by Xrayez** [comment]
I have created an initial implementation of a graph, see goostengine/goost#172.

> An example from the Python world would be NetworkX

I took some inspiration from it, and adapted to some naming conventions in Godot.

I think it's time to actually solve some concrete problem with current implementation, and see how it works out.

**Comment by Xrayez** [comment]
Implemented Kruskal's algorithm for finding minimum spanning tree in goostengine/goost#172:

![image](https://user-images.githubusercontent.com/17108460/153493113-700e41c9-ba24-4d3e-af44-b93056cc9572.png)

Or a *maximum* spanning tree, all you have to do is to inverse the weights for each edge:

![image](https://user-images.githubusercontent.com/17108460/153493459-4c88c682-9c45-486d-953c-ddd0912e94eb.png)

The performance is acceptable for moderate number of edges (takes 100 msec given 10000 edges).

For the particular case of generating [Euclidean minimum spanning tree](https://en.wikipedia.org/wiki/Euclidean_minimum_spanning_tree) (like above), this could be improved with [Delaunay triangulation](https://en.wikipedia.org/wiki/Delaunay_triangulation) for general case. I've seen https://github.com/rakai93/godot_voronoi module that allows to do something like this, and apparently uses a graph structure as well. So at some point it would be probably better to find a way to do Delaunay triangulation while retaining the structure, like in https://github.com/hiulit/Delaunator-GDScript.

Note that I had to fork Godot's `DisjointSet` data structure because it didn't have the essential `find()` operation exposed, unfortunately.

Also added serialization support for `Graph`, so you can save and load graphs as resources.

I'd say the current implementation is quite usable already.

**Comment by Zireael07** [comment]
@Xrayez: I have a GDScript delaunay implementation around, but no idea how performant my shoddy job is.

One general graph thing I would request is finding loops. Python NetworkX library does it and I always planned to implement it using NetworkX as reference, for my procedural city, but never got around to it ;)

**Comment by fire** [comment]
I recommend using the delaunay triangulation in godot math.

1. https://github.com/godotengine/godot/blob/master/core/math/delaunay_3d.h
2. https://github.com/godotengine/godot/blob/master/core/math/delaunay_2d.h

**Comment by Xrayez** [comment]
> One general graph thing I would request is finding loops.

I'm sure there has to be a way to do this with current API. The solution depends on whether you'd like to:
- detect that there's *any* cycle in the graph
- prevent cycles to appear in the first place
- find all cycles

Note that trees should not contain cycles by definition. Therefore, you can run `minimum_spanning_tree()` (with or without weights), and check edges that are present in the original graph, but *not* in the tree. If you add any such edge back to the original, it will produce a cycle. The actual process of extracting cycles is another task, and appears to be NP-hard. 😃

> I recommend using the delaunay triangulation in godot math.

Note that there exist the geometric Delaunay triangulation I've exposed previously in godotengine/godot#29127. I think the output of this could be converted to a `Graph` in `Geometry2D/3D` singletons in order to retain the structure and allow for further graph processing.

**Comment by Zireael07** [comment]
I think of the three options you mentioned,
> find or iterate through all cycles 

is the best since it allows achieving 1 and 2 also (if there's more than 1 cycle found, you have a cycle)

Good point about trees not containing cycles, though - I might try that trick.

**Comment by Xrayez** [comment]
Implemented Dijsktra's algorithm in goostengine/goost#172:

![image](https://user-images.githubusercontent.com/17108460/153933440-9748e310-35b4-4565-98d6-1f8f68e2c02e.png)

This actually constructs an entire [shortest path tree](https://en.wikipedia.org/wiki/Shortest-path_tree), from which you can collect *all* shortest path starting from any source root:

![image](https://user-images.githubusercontent.com/17108460/153933559-6b61b36d-bdf9-4810-bbb3-603fdb7ad34c.png)

The main use case is certainly procedural generation and simple path finding. I think this won't be efficient to use compared to `AStar` class in Godot (haven't tested), but the performance is good enough: 30 msec for 10000 edges. This *may* be faster in some cases since this pre-computes the entire tree.

> There's an answer at [StackOverflow which suggests to use Johnson's algorithm](https://stackoverflow.com/a/2794683/2564620), which [NetworkX appears to implement](https://stackoverflow.com/a/33956957/2564620).

[Jonhson's algorithm](https://en.wikipedia.org/wiki/Johnson%27s_algorithm) uses Dijsktra's algorithm as a subroutine, so I guess this could eventually help to implement a method to find cycles in the graph.

> > find or iterate through all cycles
> 
> is the best since it allows achieving 1 and 2 also (if there's more than 1 cycle found, you have a cycle)

but again, this doesn't seem to be a trivial problem to solve, akin to finding [longest path in the graph](https://en.wikipedia.org/wiki/Longest_path_problem) which is NP-hard. I may be wrong, though. I found a snippet if someone would be willing to implement: https://github.com/qpwo/python-simple-cycles/blob/master/johnson.py

**Comment by fire** [comment]
While working on manifold I found https://github.com/haasdo95/graphlite.

## References

https://github.com/elalish/manifold/issues/91

**Comment by m21-cerutti** [comment]
Since now goost isn't maintained anymore for godot4, would need either extract goost graph into a module or extension, create one from scratch, or make a wrapper from an other library.

Is there someone who have tried the last solution ? I have seen librairies like BGL, Melon that could be good candidates for a more general purpose graph (I would also be curious about a benchmark beetween this solutions and the legacy AStar)
https://www.boost.org/doc/libs/1_82_0/libs/graph/doc/index.html
https://github.com/fhamonic/melon

**Comment by nirahiel** [comment]
Tried to make a pure GDScript version of the goost graph for godot 4 but it's complicated. the source being in C++, there's a lot of things I don't really understand. I'm still searching online if there's something someone made already but, well, no luck so far. I guess i'll have to implement that code myself. (painfully ...)

**Comment by Raijinili** [comment]
(Based on [my post on Reddit](https://old.reddit.com/r/godot/comments/1fhkpk1/godot_should_ditch_the_astar_classes_and_just/mplqjy0/):)

You actually don't need a graph OBJECT. For many graph algorithms, all you need is:

- Type args: Node type.
- Value args: A "getNeighbors" function which, given a node, returns a list of nodes which are its neighbors.

This is because many graph algorithms work locally: At any stage, you only need to know the local structure of the graph, not the entire graph. You also don't care WHAT the node type is, as long as you can put them in a hash table, so your nodes could be numbers, or coordinates, or strings for names.

This means you don't need to ever generate a concrete graph. You don't need to create lists of edges if you can calculate it dynamically (e.g. a grid would look around for x/y adjacencies). And if you do have the edges stored, it's trivial to create the `getNeighbors` function.

This also means that you can put all the power into the `getNeighbors` function, and the graph algorithm doesn't need to understand any of that power. For example, given a game map and different unit types, a flying unit sees a different movement graph than a walking unit. If you want to do pathing for the walking unit, you don't need to generate the entire graph. You just make a closured `getNeighbors` function which refers to the actual map and the unit, and filters out any tile which the unit can't move to.

You can even do strange things like Knight's Moves and Wormholes, or non-Euclidean geometry, on a large scale, without having to encode any more of the graph than you need for the current calculation.

Limitations:
- For an A-star search, you need to be able to get neighbors AND to get the costs of moving to those neighbors. This can be given as a `getNeighborsAndCosts(v)` function (which can be lazy, and the result can be constrained to be sorted), or it can be a `getNeighbors(v)` function with a separate `costOf(u,v)` function.
- For multiedge graphs, you need to give the algorithm the specific edge you use to get to a neighbor, through a `getNeighborsByEdge` function.
- Users might not find it easy to adapt to this. It's not exactly difficult, but it isn't the first way they'd think through the problem.

----

Implementation example for, "Calculate the reachable squares, then move to one."

```python3
from typing import TypeVar, Callable, Iterator
Node = TypeVar("Node")
ReachMap = dict[Node, Node]
NeighborGetter = Callable[Node, Iterator[Node]]

def getReachables(start: Node, steps: int, getNeighbors: NeighborGetter) -> ReachMap:
    """Given a starting point, a number of steps,
    and a way to find neighbors, get a dict which
    tells you how to get from the starting point
    to another node within X steps.
    """
    reach: dict[Node, Node] = {start: None}
    candidates: list[Node] = [start]
    for i in range(steps):
        nextCandidates: list[Node] = []
        for node in candidates:
            for neighbor in getNeighbors(node):
                if neighbor in reach:  # already visited
                    continue
                reach[neighbor] = node
                nextCandidates.append(neighbor)
        candidates = nextCandidates
    return reach

def getPath(reach: ReachMap, start: Node, goal: Node) -> list[Node]:
    """Find a way to get from `start` to `goal`, given an allowed set of edges.
    
    Takes `reach` from `getReachables`.
    
    Returns a list: [start, ..., goal]
    (This can alternatively be defined as not including `start`.)
    
    Raises KeyError if `goal` can't be reached.
    """
    path = [goal]
    curNode = goal
    while curNode != start:
        curNode = reach[curNode]
        path.append(curNode)
    return list(reversed(path))
```

Usage:
```python3
# Nodes will be (int, int).
# Very large grid:
xsize, ysize = 123456789101112, 123456789101112

# Define the edges on this graph.
def getNeighbors(v):
    x, y = v
    for xd, yd in ((-1, 0), (0, -1), (1, 0), (0, 1)):
        x1 = x + xd
        y1 = y + yd
        if x1 in range(xsize) and y1 in range(ysize):
            yield (x1, y1)

# Unit's movement stat and starting point.
moveStat = 7
start = (5, 3)

# Generate the `reach` mapping.
reach = getReachables(start, moveStat, getNeighbors)

# List all reachable nodes.
reachedNodes = set(reach)

# The player chooses a node to move to.
goal = (3, 6)

# Find a path to a specific node.
path = getPath(reach, start, goal)
# [(5, 3), (4, 3), (3, 3), (3, 4), (3, 5), (3, 6)]
assert len(path) - 1 <= moveStat, "Path is too long."
assert path[0] == start, "Path didn't start at the right node."
assert path[-1] == goal, "Path didn't end at the right node."
assert all(v in set(getNeighbors(u)) for u, v in zip(path, path[1:])), "Path isn't pairwise neighbors."
```
